### PR TITLE
po/imgid - large code clean-up

### PR DESCRIPTION
**Pull request by TurboGit**
Use `dt_imgid_t` instead of gint, guint, int, int32_t or uint32_t.
Check valid imgid value using `dt_is_valid_imgid()`.
Use `NO_IMGID` to define an invalid image-id.

~~TODO: Do the same for group_id.~~

### CODE REVIEW

**Comment by jenshannoschwalm** [comment]
WHOW ! That was fast! Will check it carefully asap, right now too tired from work

File: src/common/darktable.c [hunk 1]
```diff
@@ -232,7 +232,9 @@ gboolean dt_supported_image(const gchar *filename)
   return supported;
 }
 
-int dt_load_from_string(const gchar *input, gboolean open_image_in_dr, gboolean *single_image)
+int dt_load_from_string(const gchar *input,
+                        const gboolean open_image_in_dr,
+                        gboolean *single_image)
 {
   int32_t id = 0;
```
Comment (jenshannoschwalm) [inline src/common/darktable.c:239]: Not sure about id here

File: src/common/history.c [hunk 2]
```diff
@@ -522,7 +522,7 @@ gboolean dt_history_merge_module_into_history(dt_develop_t *dev_dest,
   return module_added;
 }
 
-static gboolean _history_copy_and_paste_on_image_merge(const int32_t imgid,
+static gboolean _history_copy_and_paste_on_image_merge(const dt_imgid_t imgid,
                                                   const int32_t dest_imgid,
```
Comment (jenshannoschwalm) [inline src/common/history.c:526]: dest_imgid too

File: src/common/history.c [hunk 3]
```diff
@@ -645,7 +645,7 @@ static gboolean _history_copy_and_paste_on_image_merge(const int32_t imgid,
   return FALSE;
 }
 
-static gboolean _history_copy_and_paste_on_image_overwrite(const int32_t imgid,
+static gboolean _history_copy_and_paste_on_image_overwrite(const dt_imgid_t imgid,
                                                       const int32_t dest_imgid,
```
Comment (jenshannoschwalm) [inline src/common/history.c:649]: same here

File: src/common/history.c [hunk 4]
```diff
@@ -813,7 +813,7 @@ static gboolean _history_copy_and_paste_on_image_overwrite(const int32_t imgid,
   return ret_val;
 }
 
-gboolean dt_history_copy_and_paste_on_image(const int32_t imgid,
+gboolean dt_history_copy_and_paste_on_image(const dt_imgid_t imgid,
                                             const int32_t dest_imgid,
```
Comment (jenshannoschwalm) [inline src/common/history.c:817]: same here

File: src/common/history.c [hunk 5]
```diff
@@ -1261,7 +1261,7 @@ int dt_history_compress_on_list(const GList *imgs)
   // Get the list of selected images
   for(const GList *l = imgs; l; l = g_list_next(l))
   {
-    const int imgid = GPOINTER_TO_INT(l->data);
+    const dt_imgid_t imgid = GPOINTER_TO_INT(l->data);
```
Comment (jenshannoschwalm) [inline src/common/history.c:1264]: This is something i was wandering about.

File: src/common/image.c [hunk 6]
```diff
@@ -83,11 +83,11 @@ typedef struct dt_undo_duplicate_t
   int32_t new_imgid;
```
Comment (jenshannoschwalm) [inline src/common/image.c:83]: here too

File: src/common/image.c [hunk 7]
```diff
@@ -799,7 +799,7 @@ void dt_image_set_images_locations(const GList *imgs,
   DT_DEBUG_CONTROL_SIGNAL_RAISE(darktable.signals, DT_SIGNAL_MOUSE_OVER_IMAGE_CHANGE);
 }
 
-void dt_image_update_final_size(const int32_t imgid)
+void dt_image_update_final_size(const dt_imgid_t imgid)
 {
   if(imgid <= 0) return;
```
Comment (jenshannoschwalm) [inline src/common/image.c:804]: the macro?

File: src/common/image.c [hunk 8]
```diff
@@ -1141,12 +1141,12 @@ float dt_image_set_aspect_ratio(const int32_t imgid, const gboolean raise)
...
 }
 
-int32_t dt_image_duplicate(const int32_t imgid)
+int32_t dt_image_duplicate(const dt_imgid_t imgid)
 {
   return dt_image_duplicate_with_version(imgid, -1);
 }
 
-static int32_t _image_duplicate_with_version_ext(const int32_t imgid,
+static int32_t _image_duplicate_with_version_ext(const dt_imgid_t imgid,
```
Comment (jenshannoschwalm) [inline src/common/image.c:1149]: shouldn't these also return `dt_imgid_t`

File: src/common/image.c [hunk 9]
```diff
@@ -1345,7 +1345,7 @@ static int32_t _image_duplicate_with_version_ext(const int32_t imgid,
   return newid;
 }
 
-static int32_t _image_duplicate_with_version(const int32_t imgid,
+static int32_t _image_duplicate_with_version(const dt_imgid_t imgid,
```
Comment (jenshannoschwalm) [inline src/common/image.c:1348]: as above

File: src/common/image.c [hunk 10]
```diff
@@ -1386,12 +1386,12 @@ static int32_t _image_duplicate_with_version(const int32_t imgid,
   return newid;
 }
 
-int32_t dt_image_duplicate_with_version(const int32_t imgid, const int32_t newversion)
+int32_t dt_image_duplicate_with_version(const dt_imgid_t imgid, const int32_t newversion)
```
Comment (jenshannoschwalm) [inline src/common/image.c:1389]: as above

File: src/common/image.c [hunk 11]
```diff
@@ -2784,7 +2784,7 @@ void dt_image_synch_xmp(const int32_t selected)
 
 void dt_image_synch_all_xmp(const gchar *pathname)
 {
-  const int32_t imgid = dt_image_get_id_full_path(pathname);
+  const dt_imgid_t imgid = dt_image_get_id_full_path(pathname);
   if(imgid != -1)
```
Comment (jenshannoschwalm) [inline src/common/image.c:2788]: macro ?

File: src/common/image_cache.c [hunk 12]
```diff
@@ -203,7 +203,7 @@ void dt_image_cache_print(dt_image_cache_t *cache)
            (float)cache->cache.cost / (float)cache->cache.cost_quota);
 }
 
-dt_image_t *dt_image_cache_get(dt_image_cache_t *cache, const int32_t imgid, char mode)
+dt_image_t *dt_image_cache_get(dt_image_cache_t *cache, const dt_imgid_t imgid, char mode)
 {
   if(imgid <= 0) return NULL;
```
Comment (jenshannoschwalm) [inline src/common/image_cache.c:208]: This test is done several times, maybe the macro ?

File: src/common/metadata.c [hunk 13]
```diff
@@ -628,9 +628,9 @@ void dt_metadata_set(const int imgid, const char *key, const char *value, const
   }
 }
 
-void dt_metadata_set_import(const int imgid, const char *key, const char *value)
+void dt_metadata_set_import(const dt_imgid_t imgid, const char *key, const char *value)
 {
-  if(!key || !imgid || imgid == -1) return;
+  if(!key || !imgid || !dt_is_valid_imgid(imgid)) return;
```
Comment (jenshannoschwalm) [inline src/common/metadata.c:633]: One test for id should be enough

**Comment by TurboGit** [comment]
Thanks for review I'll fix tomorrow now. I have just pushed a new version handling group_id. This may be disruptive as there is quite a mess with testing -1, I'm not sure to have captured all cases. Anyway we're going in the right direction.

File: src/common/colorlabels.c [hunk 14]
```diff
@@ -244,7 +244,7 @@ void dt_colorlabels_toggle_label_on_list(const GList *list, const int color, con
   dt_collection_hint_message(darktable.collection);
 }
 
-int dt_colorlabels_check_label(const int imgid, const int color)
+int dt_colorlabels_check_label(const dt_imgid_t imgid, const int color)
 {
   if(imgid <= 0) return 0;
```
Comment (jenshannoschwalm) [inline src/common/colorlabels.c:249]: macro ?

File: src/common/darktable.c (hunk 1 above)
Comment (TurboGit) [reply src/common/darktable.c:239]: No as id is also used for a film.

File: src/libs/metadata_view.c [hunk 15]
```diff
@@ -994,8 +994,8 @@ static void _metadata_view_update_values(dt_lib_module_t *self)
 
 static void _jump_to()
 {
-  int32_t imgid = dt_control_get_mouse_over_id();
-  if(imgid == -1)
+  dt_imgid_t imgid = dt_control_get_mouse_over_id();
```
Comment (jenshannoschwalm) [inline src/libs/metadata_view.c:997]: const ?

File: src/libs/tools/image_infos.c [hunk 16]
```diff
@@ -85,7 +85,7 @@ void _lib_imageinfo_update_message(gpointer instance, dt_lib_module_t *self)
   dt_lib_imageinfo_t *d = (dt_lib_imageinfo_t *)self->data;
 
   // we grab the image
-  const int32_t imgid = darktable.develop->image_storage.id;
+  const dt_imgid_t imgid = darktable.develop->image_storage.id;
   if(imgid <= 0) return;
```
Comment (jenshannoschwalm) [inline src/libs/tools/image_infos.c:89]: macro

File: src/lua/gui.c [hunk 17]
```diff
@@ -379,8 +379,8 @@ static int _lua_job_valid(lua_State *L)
 
 static void _on_mouse_over_image_changed(gpointer instance, gpointer user_data)
 {
-  int imgid = dt_control_get_mouse_over_id();
-  if(imgid != -1)
+  dt_imgid_t imgid = dt_control_get_mouse_over_id();
```
Comment (jenshannoschwalm) [inline src/lua/gui.c:382]: const ?

File: src/lua/image.c [hunk 18]
```diff
@@ -357,7 +357,7 @@ static int local_copy_member(lua_State *L)
   else
   {
     dt_image_t *my_image = checkwriteimage(L, 1);
-    int imgid = my_image->id;
+    dt_imgid_t imgid = my_image->id;
```
Comment (jenshannoschwalm) [inline src/lua/image.c:360]: const ?

File: src/views/print.c [hunk 19]
```diff
@@ -321,7 +321,7 @@ gboolean try_enter(dt_view_t *self)
 
   //  now check that there is at least one selected image
 
-  const int32_t imgid = dt_act_on_get_main_image();
+  const dt_imgid_t imgid = dt_act_on_get_main_image();
 
   if(imgid < 0)
```
Comment (jenshannoschwalm) [inline src/views/print.c:326]: This seems wrong

**Review by jenshannoschwalm — COMMENTED** [review]
Couldn't resist :-) Just comments

File: src/common/image.c (hunk 11 above)
Comment (TurboGit) [reply src/common/image.c]: done

File: src/common/image_cache.c (hunk 12 above)
Comment (TurboGit) [reply src/common/image_cache.c]: done

File: src/common/image.c (hunk 9 above)
Comment (TurboGit) [reply src/common/image.c]: done

File: src/common/image.c (hunk 6 above)
Comment (TurboGit) [reply src/common/image.c]: done

File: src/common/history.c (hunk 4 above)
Comment (TurboGit) [reply src/common/history.c]: done

File: src/common/history.c (hunk 3 above)
Comment (TurboGit) [reply src/common/history.c]: done

File: src/common/history.c (hunk 2 above)
Comment (TurboGit) [reply src/common/history.c]: done

File: src/common/metadata.c (hunk 13 above)
Comment (TurboGit) [reply src/common/metadata.c]: Done.

File: src/libs/tools/image_infos.c (hunk 16 above)
Comment (TurboGit) [reply src/libs/tools/image_infos.c]: done.

File: src/lua/gui.c (hunk 17 above)
Comment (TurboGit) [reply src/lua/gui.c:382]: done.

File: src/views/print.c (hunk 19 above)
Comment (TurboGit) [reply src/views/print.c]: No, as dt_act_on_get_main_image() could return -1 before.

File: src/lua/image.c (hunk 18 above)
Comment (TurboGit) [reply src/lua/image.c:360]: done

File: src/common/colorlabels.c (hunk 14 above)
Comment (TurboGit) [reply src/common/colorlabels.c]: done.

File: src/libs/metadata_view.c (hunk 15 above)
Comment (TurboGit) [reply src/libs/metadata_view.c:997]: No, it is modified a bit later.

File: src/dtgtk/thumbtable.c [hunk 20]
```diff
@@ -1337,7 +1337,7 @@ static void _dt_mouse_over_image_callback(gpointer instance, gpointer user_data)
   if(!user_data) return;
   dt_thumbtable_t *table = (dt_thumbtable_t *)user_data;
 
-  const int imgid = dt_control_get_mouse_over_id();
+  const dt_imgid_t imgid = dt_control_get_mouse_over_id();
 
   int groupid = -1;
```
Comment (jenshannoschwalm) [inline src/dtgtk/thumbtable.c:1342]: dt_imgid here?

File: src/dtgtk/thumbtable.c [hunk 21]
```diff
@@ -2440,7 +2440,7 @@ static gboolean _zoomable_ensure_rowid_visibility(dt_thumbtable_t *table, const
   return FALSE;
 }
 
-gboolean dt_thumbtable_ensure_imgid_visibility(dt_thumbtable_t *table, const int imgid)
+gboolean dt_thumbtable_ensure_imgid_visibility(dt_thumbtable_t *table, const dt_imgid_t imgid)
 {
   if(imgid < 1) return FALSE;
```
Comment (jenshannoschwalm) [inline src/dtgtk/thumbtable.c:2445]: macro test ?

File: src/dtgtk/thumbtable.c [hunk 22]
```diff
@@ -2496,7 +2496,7 @@ static gboolean _zoomable_check_rowid_visibility(dt_thumbtable_t *table, const i
   return FALSE;
 }
 
-gboolean dt_thumbtable_check_imgid_visibility(dt_thumbtable_t *table, const int imgid)
+gboolean dt_thumbtable_check_imgid_visibility(dt_thumbtable_t *table, const dt_imgid_t imgid)
 {
   if(imgid < 1) return FALSE;
```
Comment (jenshannoschwalm) [inline src/dtgtk/thumbtable.c:2501]: here macro too?

File: src/generate-cache/main.c [hunk 23]
```diff
@@ -162,7 +162,7 @@ int main(int argc, char *arg[])
   // parse command line arguments
   dt_mipmap_size_t min_mip = DT_MIPMAP_0;
   dt_mipmap_size_t max_mip = DT_MIPMAP_2;
-  int32_t min_imgid = 0;
+  int32_t min_imgid = NO_IMGID;
```
Comment (jenshannoschwalm) [inline src/generate-cache/main.c:165]: aren't these dt_imgid_t too

File: src/libs/duplicate.c [hunk 24]
```diff
@@ -106,7 +106,7 @@ static void _lib_duplicate_duplicate_clicked_callback(GtkWidget *widget,
                                                       GdkEventButton *event,
                                                       dt_lib_module_t *self)
 {
-  const int imgid = darktable.develop->image_storage.id;
+  const dt_imgid_t imgid = darktable.develop->image_storage.id;
   const int newid = dt_image_duplicate(imgid);
```
Comment (jenshannoschwalm) [inline src/libs/duplicate.c:110]: newid is dt_imgid_t too and macro

File: src/libs/metadata_view.c [hunk 25]
```diff
@@ -994,18 +994,21 @@ static void _metadata_view_update_values(dt_lib_module_t *self)
...
```
Comment (jenshannoschwalm) [inline src/libs/metadata_view.c:997]: isnt this const here

**Review by jenshannoschwalm — COMMENTED** [review]
Some comments again

File: src/libs/metadata_view.c (hunk 25 above)
Comment (TurboGit) [reply src/libs/metadata_view.c:997]: No because modified a bit later (already reported :).

File: src/libs/duplicate.c (hunk 24 above)
Comment (TurboGit) [reply src/libs/duplicate.c]: done.

File: src/generate-cache/main.c (hunk 23 above)
Comment (TurboGit) [reply src/generate-cache/main.c]: done.

File: src/dtgtk/thumbtable.c (hunk 22 above)
Comment (TurboGit) [reply src/dtgtk/thumbtable.c]: done.

File: src/dtgtk/thumbtable.c (hunk 21 above)
Comment (TurboGit) [reply src/dtgtk/thumbtable.c]: done.

File: src/dtgtk/thumbtable.c (hunk 20 above)
Comment (TurboGit) [reply src/dtgtk/thumbtable.c]: done.

**Comment by TurboGit** [comment]
New version pushed. I'm sure we are missing some... But we will fix that probably while working/reading code later.

**Comment by TurboGit** [comment]
I have fixed some more in latest commit.

**Comment by TurboGit** [comment]
@jenshannoschwalm : If you agree and don't spot other missing parts I'll merge as this will possibly make merging other branch harder.

**Comment by jenshannoschwalm** [comment]
Go ahead, all fine for me

**Pull request merged**